sql, distsql: move projection out of sortNode by adding a parent render node

heyihong opened the pull request:
Add a parent render node if not needSort for projection and skip adding sorter in dist physical planner. Related to #20659.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20974)
<!-- Reviewable:end -->

RaduBerinde commented:
Great work, thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

CC @knz to also take a look.

I would extend the commit message to explain why we need to keep the nosort node (and add "Release note: none").

---

Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 842 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L0ugcspVPgm6iLEumUl:-L0ugcspVPgm6iLEumUm:b-50bkvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd334e6823c912c303021d0a4deb5562ffccae02/pkg/sql/distsql_physical_planner.go#L842)):*
> ```Go
> func (dsp *DistSQLPlanner) addSorters(p *physicalPlan, n *sortNode) {
> 	if !n.needSort {
> 		return
> ```

Maybe mention "This node exists only to keep track of an ordering requirement, we don't need to do anything".

---

*[pkg/sql/render.go, line 451 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L0ugD5rCm8qsQXDQeDr:-L0ugD5rCm8qsQXDQeDs:b-n4k7ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd334e6823c912c303021d0a4deb5562ffccae02/pkg/sql/render.go#L451)):*
> ```Go
> }
> 
> func (p *planner) makePostRender(ctx context.Context, src planDataSource) (*renderNode, error) {
> ```

[nit] Please add a comment for this function.

---

*[pkg/sql/render.go, line 457 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L0ugHEQ_0c252rIjpO9:-L0ugHEQ_0c252rIjpOA:b-z69qwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd334e6823c912c303021d0a4deb5562ffccae02/pkg/sql/render.go#L457)):*
> ```Go
> 	}
> 	postRender.ivarHelper = tree.MakeIndexedVarHelper(postRender, len(src.info.sourceColumns))
> 	if err := p.initTargets(ctx, postRender,
> ```

[nit] This is more readable:
```go
if err := p.initTargets(
  ctx, postRender,
  tree.SelectExprs{...},
  nil /* desiredTypes */
); err != nil
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L0uhPr4q1yVO53C-XQN:bj4rvkw)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 842 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L0ugcspVPgm6iLEumUl:-L0vBf1gOUa5DYkXMFuC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd334e6823c912c303021d0a4deb5562ffccae02/pkg/sql/distsql_physical_planner.go#L842)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Maybe mention "This node exists only to keep track of an ordering requirement, we don't need to do anything".
</blockquote></details>

Done.

---

*[pkg/sql/render.go, line 451 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L0ugD5rCm8qsQXDQeDr:-L0wmBKzOyPbyKHjtm9i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd334e6823c912c303021d0a4deb5562ffccae02/pkg/sql/render.go#L451)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Please add a comment for this function.
</blockquote></details>

Done.

---

*[pkg/sql/render.go, line 457 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L0ugHEQ_0c252rIjpO9:-L0vB39SIz8DFVInemF6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd334e6823c912c303021d0a4deb5562ffccae02/pkg/sql/render.go#L457)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] This is more readable:
```go
if err := p.initTargets(
  ctx, postRender,
  tree.SelectExprs{...},
  nil /* desiredTypes */
); err != nil
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L0yBuXfLgWvUMFZ1zZf:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Very nice optimization! Thank you for your change!

I detected a redundant function in your patch, see my comment below.
Also you will need to rebase, and re-generate the correct output for the logic tests, because we have changed the format. Thank you again for your help!

---

Reviewed 4 of 10 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/render.go, line 453 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L1NEiEuR4Wn0ORjnAcG:-L1NEiEv7zn1m4z2X_-I:b-hla9l9) ([raw file](https://github.com/cockroachdb/cockroach/blob/40e9b67c5bab1f4116d034153f531d72a0f46de9/pkg/sql/render.go#L453)):*
> ```Go
> // makeIdentityRender creates a new renderNode which selects all columns
> // in its source.
> func (p *planner) makeIdentityRender(ctx context.Context, src planDataSource) (*renderNode, error) {
> ```

Remove this and use `insertRender()` already defined above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L1NEsZmQKi4sknsdd0B:b8q8oxp)*
<!-- Sent from Reviewable.io -->

heyihong commented:
I have updated my change and rebased the code. Please take another look.

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - although

@RaduBerinde, just to verify, why do we need to *add* a renderNode on top of the noSort node if the columns match? I get why we want to keep noSort, and add a renderNode if the sort node had added some sorting columns, but if it didn't I don't see how the extra renderNode helps. Also given how it's being added here it won't be elided by the render elimination code.

---

Reviewed 12 of 12 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L2-rnhXhLQwNI1tXRqL:bagvi2q)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I thought we only add it when the columns don't match (i.e. when we need to project away some order by columns)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L20e_x1I8b7dB-oagrW:b-dq9lcx)*
<!-- Sent from Reviewable.io -->

knz commented:
Ah, yes, I misread the tests.

It is annoying for test readability that noSort doesn't announce its *real* output columns (including those that are eliminated by the new `renderNode`). @heyihong was there any reason not to include the sorting columns in the output of `planColumns(sortNode)`?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L20m0ZRxCT2MeyKPs5D:bsbulxv)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
That indeed sounds like an oversight.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L20oBDMgyHyLxocIV8r:b-lwrnz0)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Thanks for the review. The reason I used n.columns is that I thought n.columns and [planColumns(plan)](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/plan_columns.go#L69) have the same result.

heyihong commented:
Thanks for the review. The reason I used n.columns is that I thought n.columns and [planColumns(plan)](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/plan_columns.go#L69) have the same result.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/render.go, line 453 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L1NEiEuR4Wn0ORjnAcG:-L1RNDEoNyLn8UQVlWfr:brt036g) ([raw file](https://github.com/cockroachdb/cockroach/blob/40e9b67c5bab1f4116d034153f531d72a0f46de9/pkg/sql/render.go#L453)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Remove this and use `insertRender()` already defined above.
</blockquote></details>

Done. Thanks for pointing out. I also added an extra argument for specifying source columns in insertRender() so that the similar render node construction in distinct.go can reuse it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L212qZt6rnzVTonDAqq:b-g9n2xu)*
<!-- Sent from Reviewable.io -->

knz commented:
@heyihong the question is not about using `n.columns` vs `planColumns(plan)` (although, in general, we'd recommend using `planColumns`) -- the question is why does `EXPLAIN(VERBOSE)` say that the noSort node has only column `x` as output, when it also computes column `y` for `ORDER BY y`.

heyihong commented:
@knz, I think the reason is since nosort node doesn't need to sort, columns used for sorting could be optimized by not outputting them. The optimization should be executed in [setNeededColumns](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/optimize.go#L53).

RaduBerinde commented:
It would be a very minor optimization; the datums are readily available to the sortNode, it's not like it's calculating something for those columns. It doesn't cost much to just forward them and have the `renderNode` drop them. Plus most queries with sorting use distsql, and there it would make no difference (we "coalesce" projection in the sort processor regardless).

heyihong commented:
@RaduBerinde, so what I should do is to let the output columns of sortnode output equal to the output columns of sortnode.plan for readability in EXPLAIN. I am still worried about if I set n.columns to planColumns(n.plan), will it still be optimized by second-round setNeededColumns or violate some thing set by first-round setNeededColumns?

RaduBerinde commented:
`setNeededColumns` doesn't seem to depend on n.columns (other than the `markOmitted(n.columns, sourceNeeded[:len(n.columns)])` call at the end which I think should go away). I would remove `n.columns` from `sortNode` altogether.

heyihong commented:
@RaduBerinde, thanks for your explanation. I think about it for a while but I am still a little confused about how to change it:
1. if we are going to remove `n.columns` from `sortNode` altogether, how will [elideDoubleSort](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/expand_plan.go#L264-L273) work? One way I could find is to keep an array of columns name.
2. if we let the output columns of sortNode are always the same as sortNode.plan, will sortnode return redundant columns when needsort? Should we let it happen only when not needsort. For this testcase, what is the expected result after changed:
```sql
statement ok
CREATE TABLE t (
  a INT PRIMARY KEY,
  b INT,
  c BOOLEAN
)

query TITTTTT
EXPLAIN(METADATA) SELECT b FROM t ORDER BY a
----
nosort          0  nosort  ·      ·          (b)  ·
 │              0  ·       order  +a         ·    ·
 └── render     1  render  ·      ·          (b)  ·
      └── scan  2  scan    ·      ·          (b)  ·
·               2  ·       table  t@primary  ·    ·
·               2  ·       spans  ALL        ·    ·

query TITTTTT
EXPLAIN(METADATA) SELECT a FROM t ORDER BY b
----
sort            0  sort    ·      ·          (a)  a!=NULL; key(a)
 │              0  ·       order  +b         ·    ·
 └── render     1  render  ·      ·          (a)  a!=NULL; key(a)
      └── scan  2  scan    ·      ·          (a)  a!=NULL; key(a)
·               2  ·       table  t@primary  ·    ·
·               2  ·       spans  ALL        ·    ·
```

RaduBerinde commented:
1. I see, we can keep it then.

2. The whole point of this issue is to remove the projection from the sortNode; so yes, the sortNode will return some unnecessary results that the renderNode will drop. This won't matter much performance-wise, since those datums are readily available to the sortNode (and in distsql, both cases actually result in the same distsql plan).

heyihong commented:
Oh, I found that removing `n.columns` from `sortNode`may be a better way. In [p.orderBy](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sort.go#L57), we could add an extra renderNode on top of the sortNode:
```
renderNode(origin columns of sourcePlan)
    |
    |
    v
sortNode
    |
    |
    v
sourcePlan
```
In this case, we don't need to change the logic in simplifyOrderings. Then, in order to elide double sort, we could identify sortNode -> renderNode -> sortNode pattern and elide the bottom one in [doExpandPlan](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/expand_plan.go#L57). Finally, removing all codes related to `n.columns` of `sortNode`. @RaduBerinde, is it ok to do that?

RaduBerinde commented:
Yes, isn't adding a renderNode on top when we need a projection the entire point of this PR?

heyihong commented:
Sorry, I had a misunderstanding before. I only focused on TODO and thought that render node is needed when no needsort.

heyihong commented:
I refactored the change, please take another look. Current change elides sortNode->sortNode but not renderNode->sortNode->renderNode->sortNode.

knz commented:
Thanks @heyihong for your changes. However this raises a question where I'd like some input from Radu. See comments below.

---

Reviewed 16 of 17 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 863 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zto8DvnsKjWDPkxuW:-L2zto8EI0UVIN-Zwh1B:b3eyz7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/distsql_physical_planner.go#L863)):*
> ```Go
> 	if !n.needSort {
> 		// This node exists only to keep track of an ordering requirement,
> 		// we don't need to do anything
> ```

Add a period (.) at the end of the sentence.

---

*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L2zuP16fsoGmFlZOhSq:b-n8jp1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
> ```
>       └── sort            2  sort      ·            ·            (a, c)  a!=NULL; c!=NULL; key(a); +a
>            │              2  ·         order        +a,-c,+b     ·       ·
>            └── render     3  render    ·            ·            (a, c)  a!=NULL; c!=NULL; key(a); +a
> ```

@RaduBerinde just out of curiosity, why does the render node here report only columns `(a, c)` in its output yet has 3 renders? 

The contributor is probably struggling with the render optimization because the render node added on top above has the same reported output columns as the sort node. However here of course this information is incorrect, because the sort node *really* has 3 output columns not 2, so the render is actually doing something (instead of nothing) and thus should not be candidate for elimination.

---

*[pkg/sql/logictest/testdata/logic_test/order_by, line 106 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuZaIYLFEeayVJANm:-L2zuZaJ6ClaJqBZi2nB:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/order_by#L106)):*
> ```
>                 │              3  ·         order     +b         ·    ·
>                 │              3  ·         strategy  iterative  ·    ·
>                 └── render     4  render    ·         ·          (c)  weak-key(c)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L2zuxyhhzXDBC0EhTYK:b-l40e2a)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L2zwIy3BVMisNSRO5GI:b-oni0ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

@RaduBerinde just out of curiosity, why does the render node here report only columns `(a, c)` in its output yet has 3 renders? 

The contributor is probably struggling with the render optimization because the render node added on top above has the same reported output columns as the sort node. However here of course this information is incorrect, because the sort node *really* has 3 output columns not 2, so the render is actually doing something (instead of nothing) and thus should not be candidate for elimination.
</blockquote></details>

First, I don't understand why the lower levels report `key(a)`; this is only correct at the distinct level. This seems like a bug. The bug probably happens fairly late in the planning process, because if we thought that `a` was a key from the get-go, we wouldn't sort at all.

Otherwise, what could be going on here (suggested by the comment) is that we figure out that we don't need to order by `+a. -c, +b`; `+a, -c` is sufficient because the end result (after `DISTINCT(a)`) is the same. What's confusing (and probably the reason for the unnecessary renderNode ion top) is that we still keep `+a,-c,+b` in the sort node.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 9 of 22 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 863 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zto8DvnsKjWDPkxuW:-L34qqbUWa77qMk3ubjF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/distsql_physical_planner.go#L863)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Add a period (.) at the end of the sentence.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L34pji9qtrz-qRxx4Qx:b-te2yhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

First, I don't understand why the lower levels report `key(a)`; this is only correct at the distinct level. This seems like a bug. The bug probably happens fairly late in the planning process, because if we thought that `a` was a key from the get-go, we wouldn't sort at all.

Otherwise, what could be going on here (suggested by the comment) is that we figure out that we don't need to order by `+a. -c, +b`; `+a, -c` is sufficient because the end result (after `DISTINCT(a)`) is the same. What's confusing (and probably the reason for the unnecessary renderNode ion top) is that we still keep `+a,-c,+b` in the sort node.
</blockquote></details>

Thanks for pointing out the issue and explanations. Though I am still confused about why the lower levels report `key(a)`, I think I have fixed the order by issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 8 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt_needed.go, line 150 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34v_eiV6VKlf6niva9:-L34v_ejrZwKXyBoAwFC:b-u7hr63) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L150)):*
> ```Go
> 		containsNeeded := true
> 		for colIdx := range n.needed {
> 			if n.needed[colIdx] && !orderingCols.Contains(colIdx) {
> ```

Why do we care if the ordering contains all the needed columns? I would think that we care if the ordering contains a unneeded column..

---

*[pkg/sql/opt_needed.go, line 156 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34wG8WQS1TPuUjYrqH:-L34wG8WQS1TPuUjYrqI:b2af0vr) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L156)):*
> ```Go
> 		}
> 		if containsNeeded {
> 			for idx := len(n.ordering) - 1; idx != -1; idx-- {
> ```

I don't get what we're doing here. We seem to stop at the last column which is needed, but there may be more unneeded columns between that and idx = 0.. What if the ordering is `+a, +b, +c` and we determine that this is equivalent to `+a` and we only need `a` and `c`fkey(

---

*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L34wgxL9gp-lHRHRwIh:b-31f3hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Thanks for pointing out the issue and explanations. Though I am still confused about why the lower levels report `key(a)`, I think I have fixed the order by issue.
</blockquote></details>

I will investigate the key issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L34zXP-FcqL-tx7QTTr:by3o5w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I will investigate the key issue.
</blockquote></details>

It looks like it's a bug in explain, we show the properties of the root node. This explains some of the weirdness we were seeing, where the columns weren't what we expected. I'm working on a fix, I think it would be good to rebase on top of that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L352paD5nXoLnVA4XAu:bbmrxr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It looks like it's a bug in explain, we show the properties of the root node. This explains some of the weirdness we were seeing, where the columns weren't what we expected. I'm working on a fix, I think it would be good to rebase on top of that.
</blockquote></details>

#21527

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 8 of 22 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_needed.go, line 150 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34v_eiV6VKlf6niva9:-L358yvzsgLOqaH_-uiy:bi69aam) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L150)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why do we care if the ordering contains all the needed columns? I would think that we care if the ordering contains a unneeded column..
</blockquote></details>

If there are some needed columns not included in the ordering, we may not be able to keep the order of sort node's output rows unchanged after removing any unneeded ordering column.

---

*[pkg/sql/opt_needed.go, line 156 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34wG8WQS1TPuUjYrqH:-L358wfFQftr4t40ERmx:b-2kztq) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L156)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't get what we're doing here. We seem to stop at the last column which is needed, but there may be more unneeded columns between that and idx = 0.. What if the ordering is `+a, +b, +c` and we determine that this is equivalent to `+a` and we only need `a` and `c`fkey(
</blockquote></details>

My thought is to keep the order of sort node's output rows unchanged while reducing as many ordering columns as possible. So I think we could only remove the suffix of the ordering. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reviewed 1 of 14 files at r6.
Review status: 9 of 22 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_needed.go, line 150 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34v_eiV6VKlf6niva9:-L35FHqbc3ZLmBibEH_2:b4fq9my) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L150)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

If there are some needed columns not included in the ordering, we may not be able to keep the order of sort node's output rows unchanged after removing any unneeded ordering column.
</blockquote></details>

I still don't understand the new logic.. How would we even encounter a `false` in `n.needed`? AFAICT it's initialized to all 1s and never modified.

I say we should reevaluate after #21527 if we still need this new code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 5 of 23 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/opt_needed.go, line 150 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34v_eiV6VKlf6niva9:-L39E8jOaKkaOxX4gmLn:b-4sxqpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L150)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I still don't understand the new logic.. How would we even encounter a `false` in `n.needed`? AFAICT it's initialized to all 1s and never modified.

I say we should reevaluate after #21527 if we still need this new code.
</blockquote></details>

I forgot to copy needed to n.needed. It could help to eliminate some unnecessary ordering columns. I think trimming n.needed in the first round of setNeeded is safe but I am not sure trimming n.needed in the second round of setNeeded is safe. Will it lead to inconsistent ordering physical prop after second round?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: 4 of 21 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt_needed.go, line 150 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34v_eiV6VKlf6niva9:-L3SHKbdwLL-LYbcLG_T:b-2kzy0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L150)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

I forgot to copy needed to n.needed. It could help to eliminate some unnecessary ordering columns. I think trimming n.needed in the first round of setNeeded is safe but I am not sure trimming n.needed in the second round of setNeeded is safe. Will it lead to inconsistent ordering physical prop after second round?
</blockquote></details>

Though I think it is possible to reduce unneeded columns by trimming suffix of ordering, it is not the main purpose of the issue. Maybe it is better to skip it now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

knz commented:
Thank you for your work and rebasing this patch on top of the other PR. This clarifies your code greatly.

I have a few questions remaining and the code should be good to go afterwards.

---

Reviewed 3 of 14 files at r6, 15 of 15 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/expand_plan.go, line 32 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3SYIaFBTuGVDY3B-EC:-L3SYIaGj9NIjFe2t2Qc:b-w362s5) ([raw file](https://github.com/cockroachdb/cockroach/blob/23f0ad19ba17b759aaeece2826d5076f881f85f1/pkg/sql/expand_plan.go#L32)):*
> ```Go
> func (p *planner) expandPlan(ctx context.Context, plan planNode) (planNode, error) {
> 	var err error
> 	if plan, err = doExpandPlan(ctx, p, noParams, plan); err != nil {
> ```

Can you please explain this change. The resulting `plan` object should be the one passed as argument to `simplifyOrderings` below. The original code was doing this, and with your change the return value of `doExpandPlan` is scoped to the `if`. Why is this correct?

---

*[pkg/sql/opt_needed.go, line 150 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L34v_eiV6VKlf6niva9:-L3S_EClZHyG5novwAUy:b-ougdv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/447dd4bd01db6035d4b99960d11d5fdfe6f13ead/pkg/sql/opt_needed.go#L150)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

Though I think it is possible to reduce unneeded columns by trimming suffix of ordering, it is not the main purpose of the issue. Maybe it is better to skip it now.
</blockquote></details>

I have re-reviewed your patch in its entirety and I think that the field `sortNode.needed` is not needed any more. In that case, you can remove it.

---

*[pkg/sql/opt_needed.go, line 148 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3S_hdkASEncN3MxDaz:-L3S_hdkASEncN3MxDb-:b-858rkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/23f0ad19ba17b759aaeece2826d5076f881f85f1/pkg/sql/opt_needed.go#L148)):*
> ```Go
> 		copy(sourceNeeded[:len(needed)], needed)
> 
> 		// All the ordering columns are also needed.
> ```

If `needSort` is false, then they are not needed and you can omit this loop. This will optimize the plan further when the sort is not necessary.

---

*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 342 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L2zuP16fsoGmFlZOhSp:-L3SZmcORdrvV63_LAQw:bezqdrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e8afcb78f768c7751ea7b934fac1fb8e90e24e/pkg/sql/logictest/testdata/logic_test/distinct_on#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

#21527
</blockquote></details>

Thanks for figuring it out. All clear now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L3S_oQ3jRv6AFPgTPW7:b1c8bva)*
<!-- Sent from Reviewable.io -->

heyihong committed:
sql, distsql: move projection out of sortNode by adding a parent render node

Release note (sql change): moved projection out of sortNode and fixed
incorrect needed columns propagation and mark in setNeededColumns when []bool needed is a
prefix of the result columns

heyihong commented:
Thanks for the reviews. I am still working on the code. During checking logical plans, I found a issue needed to clarify. Please take a look at #21696.

---

Review status: 15 of 21 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/expand_plan.go, line 32 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3SYIaFBTuGVDY3B-EC:-L3UO-V7DRj3AorWL1UD:b-ikmpp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/23f0ad19ba17b759aaeece2826d5076f881f85f1/pkg/sql/expand_plan.go#L32)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you please explain this change. The resulting `plan` object should be the one passed as argument to `simplifyOrderings` below. The original code was doing this, and with your change the return value of `doExpandPlan` is scoped to the `if`. Why is this correct?
</blockquote></details>

I think it doesn't change the logic since both plan and err are declared out of if statement and `:=`is not used here. So both plan and err are not belong to the scope of if statement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L3XXaXwrWrdQD46ry67:b-eby374)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/expand_plan.go, line 32 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3SYIaFBTuGVDY3B-EC:-L3Xbck_4DiZ9iI1pTGd:boyej4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/23f0ad19ba17b759aaeece2826d5076f881f85f1/pkg/sql/expand_plan.go#L32)):*
<details><summary><i>Previously, heyihong (Yihong He) wrote…</i></summary><blockquote>

I think it doesn't change the logic since both plan and err are declared out of if statement and `:=`is not used here. So both plan and err are not belong to the scope of if statement.
</blockquote></details>

Oh I see. Thanks for pointing it out. Carry on.

---

*[pkg/sql/opt_needed.go, line 127 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3XbxIGCBSMVyqgRPHd:-L3XbxIGCBSMVyqgRPHe:b-aw5e4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb7833553837efc5ef125a9a60a485ed0794f7ba/pkg/sql/opt_needed.go#L127)):*
> ```Go
> 		n.ivarHelper.Reset()
> 		for i := range n.render {
> 			if i >= len(needed) || !needed[i] {
> ```

The assertion `len(needed) == len(columns)` should hold at all times. If you think that the condition `i >= len(needed)` is needed here, there must be a bug somewhere else.

---

*[pkg/sql/opt_needed.go, line 243 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3Xc8nrcNc0RxlnNvTR:-L3Xc8nrcNc0RxlnNvTS:benpsw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb7833553837efc5ef125a9a60a485ed0794f7ba/pkg/sql/opt_needed.go#L243)):*
> ```Go
> func markOmitted(cols sqlbase.ResultColumns, needed []bool) {
> 	for i := range cols {
> 		cols[i].Omitted = i >= len(needed) || !needed[i]
> ```

Same comment as above.

---

*[pkg/sql/render.go, line 421 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3XcQdQcvG6uW8C_WAq:-L3XcQdQcvG6uW8C_WAr:b-2epu1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb7833553837efc5ef125a9a60a485ed0794f7ba/pkg/sql/render.go#L421)):*
> ```Go
> // of the source plan.
> func (p *planner) insertRender(
> 	ctx context.Context, plan planNode, tn *tree.TableName, columns sqlbase.ResultColumns,
> ```

I find the new interface of this function difficult to comprehend, and also its contract (it must be a prefix...) is difficult to enforce.

Is it not more reasonable to specify the *number* of columns to use here? with `numColumns int`...

---

*[pkg/sql/render.go, line 423 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3Xc_qtc0YuOAaDesjm:-L3Xc_qtc0YuOAaDesjn:be3s3ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb7833553837efc5ef125a9a60a485ed0794f7ba/pkg/sql/render.go#L423)):*
> ```Go
> 	ctx context.Context, plan planNode, tn *tree.TableName, columns sqlbase.ResultColumns,
> ) (*renderNode, error) {
> 	src := planDataSource{info: newSourceInfoForSingleTable(*tn, columns), plan: plan}
> ```

... and then say `planColumns(plan)[:numColumns]` here.
This way it is not possible to pass an erroneous argument to the function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

knz commented:
@RaduBerinde if you have a minute I'd like your confirmation that the changes in r8 are reasonable. I'm not sure I understand them fully.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-:-L3Xd3DxGzhFVzQy_Dij:b-ad6rss)*
<!-- Sent from Reviewable.io -->

heyihong commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt_needed.go, line 127 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3XbxIGCBSMVyqgRPHd:-L3Z2mg3Cokbb1UkYLhY:bdne8t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb7833553837efc5ef125a9a60a485ed0794f7ba/pkg/sql/opt_needed.go#L127)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The assertion `len(needed) == len(columns)` should hold at all times. If you think that the condition `i >= len(needed)` is needed here, there must be a bug somewhere else.
</blockquote></details>

The reason that len(needed) != len(columns) is because len(src.info.sourceColumns) != len(planColumns(n.plan)) for renerNode and filterNode. 

The reason that len(src.info.sourceColumns) != len(planColumns(n.plan)) is because when inserting renderNode, src.info.sourceColumns is set to a prefix of planColumns(n.plan). In the current code, distinct_on is implemented like that. We should continue to work on this pull request after clarifying whether len(src.info.sourceColumns) != len(planColumns(n.plan)) is allowed. If it is allowed,  #21696 may be a fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/expand_plan.go, line 563 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974#-L3_vzFL2K7TD4VJDNrZ:-L3_vzFMRTyJuTfvbTX5:b-tl38b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb7833553837efc5ef125a9a60a485ed0794f7ba/pkg/sql/expand_plan.go#L563)):*
> ```Go
> 		}
> 		if !n.needSort {
> 			containsOrdering := true
> ```

I still don't quite get what we're doing here and why. Needs a comment at least.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20974)*
<!-- Sent from Reviewable.io -->

knz commented:
We have removed this area of the code entirely so this approach is not relevant any more. I am not 100% sure whether the inefficiency that this PR was intending to solve has been addressed in the CBO (@RaduBerinde would know) but in any case we are not pursuing this particular approach further.
